packages/react-dom/src/__tests__/ReactServerRendering-test.internal.js

```diff
@@ -384,6 +387,99 @@ describe('ReactDOMServer', () => {
       expect(markup).toContain('hello, world');
     });
 
+    it('renders with new context API', () => {
+      const Context = React.unstable_createContext(0);
+
+      function Provider(props) {
+        return Context.provide(props.value, props.children);
+      }
+
+      function Consumer(props) {
+        return Context.consume(value => {
+          return 'Result: ' + value;
+        });
+      }
+
+      const Indirection = React.Fragment;
+
+      function App(props) {
+        return (
+          <Provider value={props.value}>
+            <Provider value={2}>
+              <Consumer />
+            </Provider>
+            <Indirection>
+              <Indirection>
+                <Consumer />
+                <Provider value={3}>
+                  <Consumer />
+                </Provider>
+              </Indirection>
+            </Indirection>
+            <Consumer />
+          </Provider>
+        );
+      }
+
+      const markup = ReactDOMServer.renderToString(<App value={1} />);
+      // Extract the numbers rendered by the consumers
+      const results = markup.match(/\d+/g).map(Number);
+      expect(results).toEqual([2, 1, 3, 1]);
+    });
+
+    it('renders context API, reentrancy', () => {
+      const Context = React.unstable_createContext(0);
+
+      function Provider(props) {
+        return Context.provide(props.value, props.children);
+      }
+
+      function Consumer(props) {
+        return Context.consume(value => {
+          return 'Result: ' + value;
+        });
+      }
+
+      let reentrantMarkup;
+      function Reentrant() {
+        reentrantMarkup = ReactDOMServer.renderToString(
+          <App value={1} reentrant={false} />,
+        );
+        return null;
+      }
+
+      const Indirection = React.Fragment;
+
+      function App(props) {
+        return (
+          <Provider value={props.value}>
+            {props.reentrant && <Reentrant />}
+            <Provider value={2}>
+              <Consumer />
+            </Provider>
+            <Indirection>
+              <Indirection>
+                <Consumer />
+                <Provider value={3}>
+                  <Consumer />
+                </Provider>
+              </Indirection>
+            </Indirection>
+            <Consumer />
+          </Provider>
+        );
+      }
+
+      const markup = ReactDOMServer.renderToString(
+        <App value={1} reentrant={true} />,
+      );
+      // Extract the numbers rendered by the consumers
+      const results = markup.match(/\d+/g).map(Number);
+      const reentrantResults = reentrantMarkup.match(/\d+/g).map(Number);
+      expect(results).toEqual([2, 1, 3, 1]);
```

Comment:
@acdlite Could I ask that why it's `[2, 1, 3, 1]` rather than `[2, 1, 3, 1, 2, 1, 3, 1]`?

I think the markup shoud be 
```
<Provider value={props.value}>
  // created by Reentrant
  <Provider value={props.value}>
    {false}
    <Provider value={2}>
      <Consumer />
    </Provider>
    <Indirection>
      <Indirection>
        <Consumer />
        <Provider value={3}>
          <Consumer />
        </Provider>
      </Indirection>
    </Indirection>
    <Consumer />
  </Provider>
  // created by APP
  <Provider value={2}>
    <Consumer />
  </Provider>
  <Indirection>
    <Indirection>
      <Consumer />
      <Provider value={3}>
        <Consumer />
      </Provider>
    </Indirection>
  </Indirection>
  <Consumer />
</Provider>;
```

If I'm wrong, please let me know, thanks!